streaming/src/main/scala/org/apache/gearpump/streaming/DAG.scala

```diff
@@ -19,32 +19,54 @@
 package org.apache.gearpump.streaming
 
 import org.apache.gearpump.partitioner.{PartitionerObject, PartitionerDescription, Partitioner}
+import org.apache.gearpump.streaming.DAG.DAGDiff
 import org.apache.gearpump.streaming.task.TaskId
 import org.apache.gearpump.util.Graph
-import upickle._
 
-import scala.collection.JavaConversions._
-
-
-case class DAG(processors : Map[ProcessorId, ProcessorDescription], graph : Graph[ProcessorId, PartitionerDescription]) extends Serializable {
+case class DAG(version: Int, processors : Map[ProcessorId, ProcessorDescription], graph : Graph[ProcessorId, PartitionerDescription]) extends Serializable {
 
   def taskCount: Int = {
     processors.foldLeft(0) { (count, task) =>
       count + task._2.parallelism
     }
   }
+
+  def tasks: List[TaskId] = {
+    processors.flatMap { pair =>
+      val (processorId, processor) = pair
+      (0 until processor.parallelism).map(TaskId(processorId, _))
+    }.toList
+  }
+
+  def diff(another: DAG): DAGDiff = {
+    val left = this.processors.keySet
+    val right = another.processors.keySet
+
+    val rightOnly = right -- left
+    val leftOnly = left -- right
+    val join = right -- rightOnly
+    
+    val modifiedProcessors = join.filter(processorId =>
+      processors(processorId) != another.processors(processorId)
+    )
+
+    DAGDiff(leftOnly.toList.sorted, rightOnly.toList.sorted, modifiedProcessors.toList.sorted)
+  }
 }
 
 object DAG {
 
-  def apply (graph : Graph[ProcessorDescription, PartitionerDescription]) : DAG = {
+  case class DAGDiff(leftOnlyProcessors: List[ProcessorId], rightOnlyProcessors: List[ProcessorId],
```

Comment:
Should we support changes to partitioner as well as Config changes to both worker and executor?